### ISSUE TITLE
KeyError

### ISSUE THREAD

**Issue by veuncent** [archesproject/arches#4825]
**Describe the bug**
I am seeing errors when editing/deleting instances. This may or may not be related to #4302, having `OVERRIDE_RESOURCE_MODEL_LOCK` set to `True`.

When indexing database (`python manage.py es index_database`):
```Indexing resource type '3D HOP Model'
Traceback (most recent call last):
  File "manage.py", line 35, in <module>
    execute_from_command_line(sys.argv)
  File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/web_root/arches/arches/management/commands/es.py", line 76, in handle
    index_database.index_db(clear_index=options['clear_index'], batch_size=options['batch_size'])
  File "/web_root/arches/arches/app/utils/index_database.py", line 25, in index_db
    index_resources(clear_index=clear_index, batch_size=batch_size)
  File "/web_root/arches/arches/app/utils/index_database.py", line 44, in index_resources
    index_resources_by_type(resource_types, clear_index=clear_index, batch_size=batch_size)
  File "/web_root/arches/arches/app/utils/index_database.py", line 77, in index_resources_by_type
    document, terms = resource.get_documents_to_index(fetchTiles=True, datatype_factory=datatype_factory, node_datatypes=node_datatypes)
  File "/web_root/arches/arches/app/models/resource.py", line 216, in get_documents_to_index
    datatype = node_datatypes[nodeid]
KeyError: u'4b9b3ae4-d12f-11e8-85df-0242ac1a0004'
```

When editing a node on an existing instance:
```
2019-05-08T15:38:05.140807576Z [2019-05-08 10:38:05,140] [arches.app.views.tile] [ERROR   ] Saving tile failed [Tile id: f9482f44-7a7c-4b69-8f0a-6263e28fc2fa] [Exception message: u'4b9b3ae4-d12f-11e8-85df-0242ac1a0004'(4b9b3ae4-d12f-11e8-85df-0242ac1a0004)] [Exception trace: Traceback (most recent call last):
2019-05-08T15:38:05.140843776Z   File "/web_root/arches/arches/app/views/tile.py", line 92, in post
2019-05-08T15:38:05.142683182Z     tile.save(request=request)
2019-05-08T15:38:05.142698482Z   File "/web_root/arches/arches/app/models/tile.py", line 281, in save
2019-05-08T15:38:05.142703982Z     self.index()
2019-05-08T15:38:05.142707882Z   File "/web_root/arches/arches/app/models/tile.py", line 332, in index
2019-05-08T15:38:05.142711882Z     Resource.objects.get(pk=self.resourceinstance_id).index()
2019-05-08T15:38:05.142716082Z   File "/web_root/arches/arches/app/models/resource.py", line 182, in index
2019-05-08T15:38:05.142719882Z     document, terms = self.get_documents_to_index(datatype_factory=datatype_factory, node_datatypes=node_datatypes)
2019-05-08T15:38:05.142723882Z   File "/web_root/arches/arches/app/models/resource.py", line 216, in get_documents_to_index
2019-05-08T15:38:05.142727882Z     datatype = node_datatypes[nodeid]
2019-05-08T15:38:05.142731582Z KeyError: u'4b9b3ae4-d12f-11e8-85df-0242ac1a0004'
2019-05-08T15:38:05.142735182Z ]
2019-05-08T15:38:05.162546146Z 192.168.64.3 - - [08/May/2019:10:38:05 -0500] "POST /tile HTTP/1.0" 500 135 "https://globaldigitalheritage.org/resource/29de7adc-f69c-11e8-8ccd-0242c0a86002" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36"
```

When deleting a record (node group?) from an instance:
```
2019-05-08T15:28:06.090141639Z [2019-05-08 10:28:05,997] [django.request] [ERROR   ] Internal Server Error: /tile
2019-05-08T15:28:06.090175039Z Traceback (most recent call last):
2019-05-08T15:28:06.090186639Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
2019-05-08T15:28:06.090190939Z     response = get_response(request)
2019-05-08T15:28:06.090194639Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
2019-05-08T15:28:06.090198539Z     response = self.process_exception_by_middleware(e, request)
2019-05-08T15:28:06.090202239Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
2019-05-08T15:28:06.090214039Z     response = wrapped_callback(request, *callback_args, **callback_kwargs)
2019-05-08T15:28:06.090216439Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
2019-05-08T15:28:06.090227339Z     return self.dispatch(request, *args, **kwargs)
2019-05-08T15:28:06.090229539Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper
2019-05-08T15:28:06.090231739Z     return bound_func(*args, **kwargs)
2019-05-08T15:28:06.090233639Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 23, in _wrapped_view
2019-05-08T15:28:06.090235739Z     return view_func(request, *args, **kwargs)
2019-05-08T15:28:06.090237639Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func
2019-05-08T15:28:06.090239839Z     return func.__get__(self, type(self))(*args2, **kwargs2)
2019-05-08T15:28:06.090241839Z   File "/web_root/ENV/local/lib/python2.7/site-packages/django/views/generic/base.py", line 88, in dispatch
2019-05-08T15:28:06.090243939Z     return handler(request, *args, **kwargs)
2019-05-08T15:28:06.090245839Z   File "/web_root/arches/arches/app/views/tile.py", line 197, in delete
2019-05-08T15:28:06.090248039Z     tile.delete(request=request)
2019-05-08T15:28:06.090250039Z   File "/web_root/arches/arches/app/models/tile.py", line 319, in delete
2019-05-08T15:28:06.090252039Z     resource.index()
2019-05-08T15:28:06.090253939Z   File "/web_root/arches/arches/app/models/resource.py", line 182, in index
2019-05-08T15:28:06.090256139Z     document, terms = self.get_documents_to_index(datatype_factory=datatype_factory, node_datatypes=node_datatypes)
2019-05-08T15:28:06.090258139Z   File "/web_root/arches/arches/app/models/resource.py", line 216, in get_documents_to_index
2019-05-08T15:28:06.090260239Z     datatype = node_datatypes[nodeid]
2019-05-08T15:28:06.090262139Z KeyError: u'4b9b3ae4-d12f-11e8-85df-0242ac1a0004'
2019-05-08T15:28:06.091635842Z 192.168.64.3 - - [08/May/2019:10:28:06 -0500] "DELETE /tile HTTP/1.0" 500 27 "https://globaldigitalheritage.org/resource/29de7adc-f69c-11e8-8ccd-0242c0a86002" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36"
```

**Screenshots**
![image](https://user-images.githubusercontent.com/7893339/57388859-5e825900-71b9-11e9-8307-c55127623cf1.png)

**Your Arches Information**

*   Version used: 4.3.3 + [customizations](https://github.com/globaldigitalheritage/arches/tree/arches_3d_version_stable)
*   Operating System and version (desktop or mobile): Ubuntu 16.04 (in Docker)
*   Browser Name and version: Chrome 
*   Link to your Arches Install (optional): https://globaldigitalheritage.org/search

**Comment by chiatt** [comment]
To be addressed by #4826

**Comment by veuncent** [comment]
For future reference:

This problem was caused by having deleted a node from a Resource Model, while there were tiles (business data) present for that node. These tiles now contained orphaned data. 

This was possible because I had `OVERRIDE_RESOURCE_MODEL_LOCK` set to true (see #4826 )

To fix I had to delete the business data from these tiles. Since the tile in my case represented a node group, thus containing business data for multiple nodes in a single tile, I had to surgically remove the orphaned data:

```
update public.tiles
set tiledata = tiledata - '4b9b3ae4-d12f-11e8-85df-0242ac1a0004';
```
Where the uuid was the ID of the deleted node in the Resource Model.


To check which rows (tiles) are affected, use this select statement:
```
SELECT *
FROM public.tiles
WHERE cast(tiledata as text) LIKE '%4b9b3ae4-d12f-11e8-85df-0242ac1a0004%';
```


Initial state of the tile (column `tiledata`):
```
{
    "4b9b3ae4-d12f-11e8-85df-0242ac1a0004": [
        {
            "url": "https://prodarchesstorage.blob.core.windows.net/arches/uploadedfiles/Morgantina3DHop_Cover.jpg",
            "name": "Morgantina3DHop_Cover.jpg",
            "size": 691974,
            "type": "image/jpeg",
            "index": 0,
            "width": 1200,
            "height": 800,
            "status": "uploaded",
            "content": "blob:https://www.globaldigitalheritage.org/20aad6f5-54ef-4828-a533-07a2aabcc10f",
            "file_id": "2798d346-18ff-11e9-af39-0242ac130002",
            "accepted": true,
            "lastModified": 1547581949791
        }
    ],
    "4b9b3c7e-d12f-11e8-85df-0242ac1a0004": [
        {
            "url": "https://prodarchesstorage.blob.core.windows.net/arches/uploadedfiles/Morgantina3DHop_Thumbnail.jpg",
            "name": "Morgantina3DHop_Thumbnail.jpg",
            "size": 102806,
            "type": "image/jpeg",
            "index": 0,
            "width": 400,
            "height": 400,
            "status": "uploaded",
            "content": "blob:https://www.globaldigitalheritage.org/16295d9d-8028-4167-af8a-019f05f6898f",
            "file_id": "278d33e2-18ff-11e9-af39-0242ac130002",
            "accepted": true,
            "lastModified": 1547581907423
        }
    ]
}
```

State of the tile (column `tiledata`) after deleting orphaned data:
```
{
    "4b9b3c7e-d12f-11e8-85df-0242ac1a0004": [
        {
            "url": "https://prodarchesstorage.blob.core.windows.net/arches/uploadedfiles/Morgantina3DHop_Thumbnail.jpg",
            "name": "Morgantina3DHop_Thumbnail.jpg",
            "size": 102806,
            "type": "image/jpeg",
            "index": 0,
            "width": 400,
            "height": 400,
            "status": "uploaded",
            "content": "blob:https://www.globaldigitalheritage.org/16295d9d-8028-4167-af8a-019f05f6898f",
            "file_id": "278d33e2-18ff-11e9-af39-0242ac130002",
            "accepted": true,
            "lastModified": 1547581907423
        }
    ]
}
```
